packages/next-swc/crates/next-api/src/dynamic_imports.rs

```diff
@@ -0,0 +1,395 @@
+    let imported_modules_mapping = NonDeterministic::new()
+        .skip_duplicates()
+        .visit([Vc::upcast(entry)], get_referenced_modules)
+        .await
+        .completed()?
+        .into_inner()
+        .into_iter()
+        .map(build_dynamic_imports_map_for_module);
+
+    // Consolidate import mappings into a single indexmap
+    let mut import_mappings: IndexMap<Vc<Box<dyn Module>>, DynamicImportedModules> =
+        IndexMap::new();
+
+    for module_mapping in imported_modules_mapping {
+        if let Some(module_mapping) = &*module_mapping.await? {
+            let (origin_module, dynamic_imports) = &*module_mapping.await?;
+            import_mappings
+                .entry(*origin_module)
+                .or_insert_with(Vec::new)
+                .append(&mut dynamic_imports.clone())
+        }
+    }
+
+    Ok(import_mappings)
+}
+
+async fn get_referenced_modules(
+    parent: Vc<Box<dyn Module>>,
+) -> Result<impl Iterator<Item = Vc<Box<dyn Module>>> + Send> {
+    primary_referenced_modules(parent)
+        .await
+        .map(|modules| modules.clone_value().into_iter())
+}
+
+#[turbo_tasks::function]
+async fn build_dynamic_imports_map_for_module(
+    module: Vc<Box<dyn Module>>,
+) -> Result<Vc<OptionDynamicImportsMap>> {
+    let Some(ecmascript_asset) =
+        Vc::try_resolve_downcast_type::<EcmascriptModuleAsset>(module).await?
+    else {
+        return Ok(OptionDynamicImportsMap::none());
+    };
+
+    let ParseResult::Ok { program, .. } = &*ecmascript_asset.parse().await? else {
+        NextDynamicParsingIssue {
+            ident: module.ident(),
+        }
+        .cell()
+        .emit();
+
+        return Ok(OptionDynamicImportsMap::none());
+    };
+
+    // Reading the Program AST, collect raw imported module str if it's wrapped in
+    // dynamic()
+    let mut visitor = DynamicImportVisitor::new();
+    program.visit_with(&mut visitor);
+
+    if visitor.import_sources.is_empty() {
+        return Ok(OptionDynamicImportsMap::none());
+    }
+
+    let mut import_sources = vec![];
+    for import in visitor.import_sources.drain(..) {
+        // Using the given `Module` which is the origin of the dynamic import, trying to
+        // resolve the module that is being imported.
+        let dynamic_imported_resolved_module = *esm_resolve(
+            Vc::upcast(PlainResolveOrigin::new(
+                ecmascript_asset.await?.asset_context,
+                module.ident().path(),
+            )),
+            Request::parse(Value::new(Pattern::Constant(import.to_string()))),
+            Value::new(EcmaScriptModulesReferenceSubType::Undefined),
+            OptionIssueSource::none(),
+            IssueSeverity::Error.cell(),
+        )
+        .first_module()
+        .await?;
+
+        if let Some(dynamic_imported_resolved_module) = dynamic_imported_resolved_module {
+            import_sources.push((import, dynamic_imported_resolved_module));
+        }
```

Comment:
If we dynamic load the same module multiple times from the same source, can we dedupe these?

Reply:
I guess not? since this is per-origin, each origin should have all the chunks pointing to the import sources to be loaded. The reason original manifest points to the chunk path is to load those as chunks, so if the imported modules are same and if other origin already loaded it can be avoided.